File: One2lvOS/Agentic_Control/ai-lobby/astraVectorMemory.js
/**
 * Astra DB Vector Memory Integration for One2lvOS
 * Node.js client for managing vector embeddings using NVIDIA embedding provider
 */

const axios = require('axios');

class AstraVectorMemory {
    constructor() {
        this.apiEndpoint = process.env.ASTRA_DB_API_ENDPOINT;
        this.token = process.env.ASTRA_DB_APPLICATION_TOKEN;
        this.keyspace = process.env.ASTRA_DB_KEYSPACE || 'sovereign_memory';
        this.model = process.env.ASTRA_NVIDIA_MODEL || 'nvidia/nv-embedqa-e5-v5';

        if (!this.apiEndpoint || !this.token) {
            throw new Error('ASTRA_DB credentials required');
        }

        this.headers = {
            'Token': this.token,
            'Content-Type': 'application/json'
        };

        this.collections = {
            agentMemory: 'agent_memory',
            council: 'sovereign_council',
            delta: 'delta_engine_state',
            snapshots: 'system_snapshots',
            interactions: 'user_interactions'
        };
    }

    /**
     * Make API call to Astra DB
     */
    async _apiCall(collection, payload) {
        const url = `${this.apiEndpoint}/api/json/v1/${this.keyspace}/${collection}`;
        const response = await axios.post(url, payload, { headers: this.headers });
        return response.data;
    }

    /**
     * Store agent memory with automatic NVIDIA embedding
     */
    async storeAgentMemory(agentName, content, metadata = {}) {
        const document = {
            agent: agentName,
            content: content,
            timestamp: new Date().toISOString(),
            $vectorize: content, // Automatic NVIDIA embedding
            ...metadata
        };

        const payload = {
            insertOne: {
                document: document
            }
        };

        const result = await this._apiCall(this.collections.agentMemory, payload);
        return result.status?.insertedIds?.[0];
    }

    /**
     * Search agent memories using vector similarity
     */
    async searchAgentMemory(query, agentName = null, limit = 5) {
        const payload = {
            find: {
                sort: {
                    $vectorize: query
                },
                options: {
                    limit: limit,
                    includeSimilarity: true
                }
            }
        };

        if (agentName) {
            payload.find.filter = { agent: agentName };
        }

        const result = await this._apiCall(this.collections.agentMemory, payload);
        return result.data?.documents || [];
    }

    /**
     * Store Sovereign Council deliberation
     */
    async storeCouncilDeliberation(decisionId, agents, topic, deliberation, consensus, votes) {
        const document = {
            decision_id: decisionId,
            agents: agents,
            topic: topic,
            deliberation: deliberation,
            consensus: consensus,
            votes: votes,
            timestamp: new Date().toISOString(),
            $vectorize: `${topic}: ${deliberation}`
        };

        const payload = {
            insertOne: {
                document: document
            }
        };

        const result = await this._apiCall(this.collections.council, payload);
        return result.status?.insertedIds?.[0];
    }

    /**
     * Search past council decisions
     */
    async searchCouncilDecisions(query, minConsensus = null, limit = 5) {
        const payload = {
            find: {
                sort: {
                    $vectorize: query
                },
                options: {
                    limit: limit,
                    includeSimilarity: true
                }
            }
        };

        if (minConsensus !== null) {
            payload.find.filter = {
                consensus: { $gte: minConsensus }
            };
        }

        const result = await this._apiCall(this.collections.council, payload);
        return result.data?.documents || [];
    }

    /**
     * Store Delta Engine state snapshot
     */
    async storeDeltaState(cycle, energy, stability, momentum, stateDescription, fullState) {
        const document = {
            cycle: cycle,
            energy: energy,
            stability: stability,
            momentum: momentum,
            state_description: stateDescription,
            full_state: fullState,
            timestamp: new Date().toISOString(),
            $vectorize: `Cycle ${cycle}: ${stateDescription}`
        };

        const payload = {
            insertOne: {
                document: document
            }
        };

        const result = await this._apiCall(this.collections.delta, payload);
        return result.status?.insertedIds?.[0];
    }

    /**
     * Store O2PNG snapshot metadata
     */
    async storeO2PNGSnapshot(snapshotId, generation, description, snapshotPath, metadata) {
        const document = {
            snapshot_id: snapshotId,
            generation: generation,
            description: description,
            snapshot_path: snapshotPath,
            metadata: metadata,
            timestamp: new Date().toISOString(),
            $vectorize: `Snapshot ${generation}: ${description}`
        };

        const payload = {
            insertOne: {
                document: document
            }
        };

        const result = await this._apiCall(this.collections.snapshots, payload);
        return result.status?.insertedIds?.[0];
    }

    /**
     * Search O2PNG snapshots by description
     */
    async searchSnapshots(query, limit = 5) {
        const payload = {
            find: {
                sort: {
                    $vectorize: query
                },
                options: {
                    limit: limit,
                    includeSimilarity: true
                }
            }
        };

        const result = await this._apiCall(this.collections.snapshots, payload);
        return result.data?.documents || [];
    }

    /**
     * Store user interaction
     */
    async storeUserInteraction(userId, command, response, context = {}) {
        const document = {
            user_id: userId,
            command: command,
            response: response,
            context: context,
            timestamp: new Date().toISOString(),
            $vectorize: `User: ${command}\nAssistant: ${response}`
        };

        const payload = {
            insertOne: {
                document: document
            }
        };

        const result = await this._apiCall(this.collections.interactions, payload);
        return result.status?.insertedIds?.[0];
    }

    /**
     * Get relevant context from all collections
     */
    async getRelevantContext(query, limit = 10) {
        const context = {};

        for (const [name, collection] of Object.entries(this.collections)) {
            try {
                const payload = {
                    find: {
                        sort: {
                            $vectorize: query
                        },
                        options: {
                            limit: limit,
                            includeSimilarity: true
                        }
                    }
                };

                const result = await this._apiCall(collection, payload);
                context[name] = result.data?.documents || [];
            } catch (error) {
                console.error(`Error fetching context from ${name}:`, error.message);
                context[name] = [];
            }
        }

        return context;
    }

    /**
     * Store AI model response with embeddings
     */
    async storeModelResponse(modelName, prompt, response, metadata = {}) {
        return this.storeAgentMemory(
            modelName,
            `Prompt: ${prompt}\nResponse: ${response}`,
            {
                type: 'model_response',
                model: modelName,
                prompt: prompt,
                response: response,
                ...metadata
            }
        );
    }

    /**
     * Search similar conversations
     */
    async searchSimilarConversations(query, modelName = null, limit = 5) {
        const memories = await this.searchAgentMemory(query, modelName, limit);
        return memories.filter(m => m.type === 'model_response');
    }
}

// Singleton instance
let astraMemoryInstance = null;

function getAstraMemory() {
    if (!astraMemoryInstance) {
        astraMemoryInstance = new AstraVectorMemory();
    }
    return astraMemoryInstance;
}

module.exports = {
    AstraVectorMemory,
    getAstraMemory
};

File: One2lvOS/Agentic_Control/ai-lobby/sovereign-council-bridge/astra_vector_memory.py
"""
Astra DB Vector Memory Integration for One2lvOS
Manages vector embeddings using NVIDIA embedding provider through Astra vectorize
"""

import os
import json
import requests
from typing import Dict, List, Optional, Any
from datetime import datetime


class AstraVectorMemory:
    """Vector memory management using Astra DB with NVIDIA embeddings"""

    def __init__(self):
        self.api_endpoint = os.getenv("ASTRA_DB_API_ENDPOINT")
        self.token = os.getenv("ASTRA_DB_APPLICATION_TOKEN")
        self.keyspace = os.getenv("ASTRA_DB_KEYSPACE", "sovereign_memory")
        self.model = os.getenv("ASTRA_NVIDIA_MODEL", "nvidia/nv-embedqa-e5-v5")

        if not self.api_endpoint or not self.token:
            raise ValueError("ASTRA_DB credentials required")

        self.headers = {
            "Token": self.token,
            "Content-Type": "application/json"
        }

        self.collections = {
            "agent_memory": "agent_memory",
            "council": "sovereign_council",
            "delta": "delta_engine_state",
            "snapshots": "system_snapshots",
            "interactions": "user_interactions"
        }

    def _api_call(self, collection: str, payload: Dict) -> Dict:
        """Make API call to Astra DB"""
        url = f"{self.api_endpoint}/api/json/v1/{self.keyspace}/{collection}"
        response = requests.post(url, headers=self.headers, json=payload)
        response.raise_for_status()
        return response.json()

    async def store_agent_memory(
        self,
        agent_name: str,
        content: str,
        metadata: Optional[Dict] = None
    ) -> str:
        """
        Store agent memory with automatic NVIDIA embedding

        Args:
            agent_name: Name of the agent
            content: Memory content to vectorize
            metadata: Additional metadata

        Returns:
            Document ID
        """
        document = {
            "agent": agent_name,
            "content": content,
            "timestamp": datetime.utcnow().isoformat(),
            "$vectorize": content,  # Automatic NVIDIA embedding
            **(metadata or {})
        }

        payload = {"insertOne": {"document": document}}
        result = self._api_call(self.collections["agent_memory"], payload)

        return result.get("status", {}).get("insertedIds", [None])[0]

    async def search_agent_memory(
        self,
        query: str,
        agent_name: Optional[str] = None,
        limit: int = 5
    ) -> List[Dict]:
        """
        Search agent memories using vector similarity

        Args:
            query: Search query (will be vectorized by NVIDIA)
            agent_name: Filter by specific agent
            limit: Number of results

        Returns:
            List of similar memories with similarity scores
        """
        payload = {
            "find": {
                "sort": {"$vectorize": query},
                "options": {
                    "limit": limit,
                    "includeSimilarity": True
                }
            }
        }

        if agent_name:
            payload["find"]["filter"] = {"agent": agent_name}

        result = self._api_call(self.collections["agent_memory"], payload)
        return result.get("data", {}).get("documents", [])

    async def store_council_deliberation(
        self,
        decision_id: str,
        agents: List[str],
        topic: str,
        deliberation: str,
        consensus: float,
        votes: Dict[str, Any]
    ) -> str:
        """Store Sovereign Council deliberation"""
        document = {
            "decision_id": decision_id,
            "agents": agents,
            "topic": topic,
            "deliberation": deliberation,
            "consensus": consensus,
            "votes": votes,
            "timestamp": datetime.utcnow().isoformat(),
            "$vectorize": f"{topic}: {deliberation}"
        }

        payload = {"insertOne": {"document": document}}
        result = self._api_call(self.collections["council"], payload)

        return result.get("status", {}).get("insertedIds", [None])[0]

    async def search_council_decisions(
        self,
        query: str,
        min_consensus: Optional[float] = None,
        limit: int = 5
    ) -> List[Dict]:
        """Search past council decisions"""
        payload = {
            "find": {
                "sort": {"$vectorize": query},
                "options": {
                    "limit": limit,
                    "includeSimilarity": True
                }
            }
        }

        if min_consensus is not None:
            payload["find"]["filter"] = {"consensus": {"$gte": min_consensus}}

        result = self._api_call(self.collections["council"], payload)
        return result.get("data", {}).get("documents", [])

    async def store_delta_state(
        self,
        cycle: int,
        energy: float,
        stability: float,
        momentum: float,
        state_description: str,
        full_state: Dict
    ) -> str:
        """Store Delta Engine state snapshot"""
        document = {
            "cycle": cycle,
            "energy": energy,
            "stability": stability,
            "momentum": momentum,
            "state_description": state_description,
            "full_state": full_state,
            "timestamp": datetime.utcnow().isoformat(),
            "$vectorize": f"Cycle {cycle}: {state_description}"
        }

        payload = {"insertOne": {"document": document}}
        result = self._api_call(self.collections["delta"], payload)

        return result.get("status", {}).get("insertedIds", [None])[0]

    async def store_o2png_snapshot(
        self,
        snapshot_id: str,
        generation: int,
        description: str,
        snapshot_path: str,
        metadata: Dict
    ) -> str:
        """Store O2PNG snapshot metadata"""
        document = {
            "snapshot_id": snapshot_id,
            "generation": generation,
            "description": description,
            "snapshot_path": snapshot_path,
            "metadata": metadata,
            "timestamp": datetime.utcnow().isoformat(),
            "$vectorize": f"Snapshot {generation}: {description}"
        }

        payload = {"insertOne": {"document": document}}
        result = self._api_call(self.collections["snapshots"], payload)

        return result.get("status", {}).get("insertedIds", [None])[0]

    async def search_snapshots(
        self,
        query: str,
        limit: int = 5
    ) -> List[Dict]:
        """Search O2PNG snapshots by description"""
        payload = {
            "find": {
                "sort": {"$vectorize": query},
                "options": {
                    "limit": limit,
                    "includeSimilarity": True
                }
            }
        }

        result = self._api_call(self.collections["snapshots"], payload)
        return result.get("data", {}).get("documents", [])

    async def store_user_interaction(
        self,
        user_id: str,
        command: str,
        response: str,
        context: Optional[Dict] = None
    ) -> str:
        """Store user interaction"""
        document = {
            "user_id": user_id,
            "command": command,
            "response": response,
            "context": context or {},
            "timestamp": datetime.utcnow().isoformat(),
            "$vectorize": f"User: {command}\nAssistant: {response}"
        }

        payload = {"insertOne": {"document": document}}
        result = self._api_call(self.collections["interactions"], payload)

        return result.get("status", {}).get("insertedIds", [None])[0]

    async def get_relevant_context(
        self,
        query: str,
        limit: int = 10
    ) -> Dict[str, List[Dict]]:
        """
        Get relevant context from all collections

        Args:
            query: Context query
            limit: Results per collection

        Returns:
            Dictionary with results from each collection
        """
        context = {}

        for name, collection in self.collections.items():
            try:
                payload = {
                    "find": {
                        "sort": {"$vectorize": query},
                        "options": {
                            "limit": limit,
                            "includeSimilarity": True
                        }
                    }
                }

                result = self._api_call(collection, payload)
                context[name] = result.get("data", {}).get("documents", [])
            except Exception as e:
                print(f"Error fetching context from {name}: {e}")
                context[name] = []

        return context


# Singleton instance
_astra_memory = None


def get_astra_memory() -> AstraVectorMemory:
    """Get or create Astra memory instance"""
    global _astra_memory
    if _astra_memory is None:
        _astra_memory = AstraVectorMemory()
    return _astra_memory
